candle-transformers/src/models/whisper/audio.rs

```diff
@@ -180,10 +212,55 @@ fn log_mel_spectrogram_<T: Float + std::fmt::Display>(
         samples_padded
     };
 
-    // Use a single thread for now.
-    let mut mel = log_mel_spectrogram_w(
-        0, &hann, &samples, filters, fft_size, fft_step, speed_up, n_len, n_mel, 1,
-    );
+    // ensure that the number of threads is even and less than 12
+    let n_threads = std::cmp::min(get_num_threads() - get_num_threads() % 2, 12);
+
+    let hann = Arc::new(hann);
+    let samples = Arc::new(samples);
+    let cloned_contents_of_filters = filters.to_vec();
+    let filters = Arc::new(cloned_contents_of_filters);
+
+    let mut handles = Vec::new();
+    for thread_id in 0..n_threads {
+        let hann_clone = hann.clone();
+        let samples_clone = samples.clone();
+        let filters_clone = filters.clone();
+
+        handles.push(thread::spawn(move || {
+            log_mel_spectrogram_w(
+                thread_id,
+                &hann_clone,
+                &samples_clone,
+                &filters_clone,
+                fft_size,
+                fft_step,
+                speed_up,
+                n_len,
+                n_mel,
+                n_threads,
+            )
+        }));
+    }
+
+    let mut all_outputs = Vec::new();
+    for handle in handles {
+        match handle.join() {
+            Ok(output) => all_outputs.push(output),
+            Err(_) => panic!("Thread failed"),
+        }
+    }
+
+    let l = all_outputs[0].len();
+    let mut mel = vec![zero; l];
+
+    for i in (0..l).step_by(n_threads) {
+        for ouputs in all_outputs.iter() {
+            for j in 0..n_threads {
```

Comment:
Could you add some comment to detail a bit what is going on here, the first loop on the outputs of all threads then iterating on the number of thtreads is pretty hard to follow.

Reply:
yea forsure, I agree the logic is a bit unintuitive (followed the pattern from llama.cpp) and I've rewritten it to be slightly more readable. (copied below for reference). 

```rust
    // iterate over mel spectrogram segments, dividing work by threads.
    for segment_start in (0..l).step_by(n_threads) {
        // go through each thread's output.
        for thread_output in all_outputs.iter() {
            // add each thread's piece to our mel spectrogram.
            for offset in 0..n_threads {
                let mel_index = segment_start + offset; // find location in mel.
                if mel_index < mel.len() {
                    // Make sure we don't go out of bounds.
                    mel[mel_index] += thread_output[mel_index];
                }
            }
        }
    }
```

in general the pattern distributes the work (chunks of the input) across the threads with different offsets and the output values for each thread is offset but that amount too. printing them before joining has the following pattern
```
[-10.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -4.4549956, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
[0.0, -10.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -2.9593704, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
[0.0, 0.0, -10.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -3.7152524, 0.0, 0.0, 0.0, 0.0, 0.0]
[0.0, 0.0, 0.0, -9.601439, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -3.808778, 0.0, 0.0, 0.0, 0.0]
[0.0, 0.0, 0.0, 0.0, -7.369365, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -3.279217, 0.0, 0.0, 0.0]
[0.0, 0.0, 0.0, 0.0, 0.0, -3.5668745, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -4.3706155, 0.0, 0.0]
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -3.502727, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -3.2597005, 0.0]
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -3.904054, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -2.861027]
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -3.7282937, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -3.0264673, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -3.0763023, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, -3.3680458, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
```